File: src/activity/NormalDistribution.java
package activity;

public class NormalDistribution {

    private double[] data;

    NormalDistribution(double[] data) {
        this.data = data;
    }



    double getMean() {
        double sum = 0;

        for(double a : data)
            sum += a;

        return sum / data.length;
    }

    double getStdDev() {
        double mean = getMean();
        double sum = 0;

        for(double a : data)
            sum += a * a;

        sum /= data.length;

        return Math.sqrt(sum - mean * mean);
    }


    private double normal(double x, double mean, double std) {
        var arg = 1d / (std * Math.sqrt(2 * Math.PI));

        var power = - ((x - mean) * (x - mean)) / (2 * std * std);

        return arg * Math.exp(power);
    }

    double normal(double x) {
        return normal(x, getMean(), getStdDev());
    }
}

File: src/activity/NormalDistribution.scala
package activity

class NormalDistribution(data: Array[Double]) {

    def getMean: Double = {
        var sum = 0d

        for (a: Double <- data) sum += a

        sum / data.length
    }

    def getStdDev: Double = {
        val mean = getMean
        var sum = 0d

        for (a <- data) sum += a * a

        sum /= data.length

        Math.sqrt(sum - mean * mean)
    }

    private def normal(x: Double, mean: Double, std: Double): Double = {
        val arg = 1d / (std * Math.sqrt(2 * Math.PI))

        val power = -(x - mean) * (x - mean) / (2 * std * std)

        arg * Math.exp(power)
    }

    private[activity] def normal(x: Double): Double = normal(x, getMean, getStdDev)
}